wfc_gh/WaveFunctionCollapse.cs

```diff
@@ -394,141 +297,190 @@ protected override void SolveInstance(IGH_DataAccess DA)
+
+                    ushort x = (ushort)xInt;
+                    ushort y = (ushort)yInt;
+                    ushort z = (ushort)zInt;
+
+                    if (nameToModule.TryGetValue(moduleStr.ToString(), out ushort m)) {
+                        var result = Native.wfc_world_state_slot_module_set(wfcWorldStateHandle, x, y, z, m, 1);
+                        switch (result) {
+                            case WfcWorldStateSlotModuleSetResult.Ok:
+                                break;
+                            case WfcWorldStateSlotModuleSetResult.ErrSlotOutOfBounds:
+                                AddRuntimeMessage(GH_RuntimeMessageLevel.Error,
+                                                  "Provided slots must be within bounds of the world");
+
+                                Native.wfc_world_state_free(wfcWorldStateHandle);
+                                return;
+                            case WfcWorldStateSlotModuleSetResult.ErrModuleOutOfBounds:
+                            default:
+                                Debug.Assert(false);
+                                return;
+                        }
+                    } else {
+                        AddRuntimeMessage(GH_RuntimeMessageLevel.Error,
+                                          "Slot module list (SM) contains module not found in the ruleset: " + moduleStr);
+
+                        Native.wfc_world_state_free(wfcWorldStateHandle);
+                        return;
+                    }
+                }
 
-                fixed (SlotState* slotsPtr = &slots[0])
                 {
-                    var result = Native.wfc_world_state_slots_set(wfcWorldStateHandle,
-                                                                  slotsPtr,
-                                                                  (UIntPtr)slots.Length);
-                    switch (result)
-                    {
-                        case WfcWorldStateSlotsSetResult.Ok:
-                            // All good
-                            stats.worldNotCanonical = false;
-                            break;
-                        case WfcWorldStateSlotsSetResult.OkWorldNotCanonical:
-                            // All good, but we the slots we gave were not
-                            // canonical. wfc_world_state_slots_set fixed that for us.
-                            stats.worldNotCanonical = true;
+                    var result = Native.wfc_world_state_canonicalize(wfcWorldStateHandle);
+                    switch (result) {
+                        case WfcWorldStateCanonicalizeResult.OkDeterministic:
+                        case WfcWorldStateCanonicalizeResult.OkNondeterministic:
+                        case WfcWorldStateCanonicalizeResult.OkContradiction:
+                            // We could potentially early out here if we are deterministic, or have a contradiction
+                            // but wfc_observe deals with that too.
```

Comment:
what if we limit the number of observations to 0?

also why are all the options ok? if there is no non-ok option, why even stating that it is ok?

Reply:
> what if we limit the number of observations to 0?

In theory, you shouldn't need to call `wfc_observe` at all in this case, but it doesn't hurt.

> also why are all the options ok? if there is no non-ok option, why even stating that it is ok?

They really all are Ok and expected. Naming is just for consistency, although I was considering returning this as an out parameter, to be even more consistent with the rest of the API, where there is just one Ok variant.

Reply:
So why not early out here? In Monoceros we can potentially skip any observation.

Reply:
We could/should early out. I just didn't out of laziness. I'll update the code.

Reply:
I just tried adding the early-out and found out it adds an unreasonable amount of code for little value. Because this is just a reference implementation and nobody really uses it, I'd leave it be as a comment.